### PR TITLE
async api

### PR DESCRIPTION
**Pull request by legokichi**
- redesign async API to use Futures and Tokio

### CODE REVIEW

File: rust-toolchain [hunk 1]
```diff
@@ -1 +1 @@
-1.49.0
```
Comment (legokichi) [inline rust-toolchain:1]: :eyes:

**Comment by garasubo** [comment]
Can we keep blocking interface as a different packages for easy migration?

**Comment by legokichi** [comment]
@garasubo https://stackoverflow.com/questions/58739075/how-do-i-import-multiple-versions-of-the-same-crate

File: Cargo.toml [hunk 2]
```diff
@@ -38,21 +41,20 @@ url = "2"
 byteorder = "1"
 tar = "0.4"
 tokio = { version = "1", features = ["full"] }
```
Comment (garasubo) [inline Cargo.toml:43]: do we still need full feature tokio?

Comment (legokichi) [reply Cargo.toml:43]: perhaps

```suggestion
tokio = { version = "1", features = ["time", "sync", "fs", "net"] }
```
https://docs.rs/crate/tokio-stream/0.1.12/features

File: src/hyper_client.rs [hunk 3]
```diff
@@ -31,174 +27,21 @@ impl Client {
...
-
-        self.buf = buffer;
-
-        Ok(i)
-    }
-}
-
 /// Http client using hyper
-#[derive(Debug)]
+#[derive(Debug, Clone)]
```
Comment (legokichi) [inline src/hyper_client.rs:31]: changed

File: src/docker.rs [hunk 4]
```diff
@@ -66,7 +153,7 @@ enum Protocol {
 }
 
 /// Handle to connection to the docker daemon
-#[derive(Debug)]
+#[derive(Debug, Clone)]
```
Comment (legokichi) [inline src/docker.rs:156]: changed

File: src/docker.rs [hunk 5]
```diff
@@ -146,12 +210,13 @@ impl Docker {
...
             protocol,
             headers: HeaderMap::new(),
-            credential: None,
+            credential: std::sync::Arc::new(std::sync::Mutex::new(None)),
         }
     }
 
-    pub fn set_credential(&mut self, credential: Credential) {
-        self.credential = Some(credential)
+    pub fn set_credential(&self, credential: Credential) {
```
Comment (legokichi) [inline src/docker.rs:217]: changed

File: src/hyper_client.rs [hunk 6]
```diff
@@ -261,12 +104,10 @@ async fn request_with_redirect<T: Into<hyper::Body> + Sync + Send + 'static + Cl
 
                 future = client.request(if see_other {
                     request.body(hyper::Body::empty()).unwrap()
+                } else if let Some(body) = body.clone() {
```
Comment (garasubo) [inline src/hyper_client.rs:107]: Nit: Is this `clone` necessary?

**Review by garasubo — APPROVED** [review]
Some nit comments, but LGTM

File: src/hyper_client.rs (hunk 6 above)
Comment (legokichi) [reply src/hyper_client.rs:107]: because the body is moved in the loop

**Pull request merged**